pkg/updatestatus/updatestatuscontroller.go

```diff
@@ -0,0 +1,234 @@
+	return strings.HasPrefix(k, "usc-")
+}
+
+// updateStatusController is a controller that collects insights from informers and maintains a ConfigMap with the insights
+// until we have a proper UpdateStatus API. The controller maintains an internal desired content of the ConfigMap (even
+// if it does not exist in the cluster) and updates it in the cluster when new insights are received, or when the ConfigMap
+// changes in the cluster. The controller only maintains the ConfigMap in the cluster if it exists, it does not create it
+// itself (this serves as a simple opt-in mechanism).
+//
+// The communication between informers (insight producers) and this controller is performed via a channel.
+// The newUpdateStatusController constructor calls the startCollectingInsights function that creates the channel spawns
+// a goroutine that consumes the channel and updates the internal ConfigMap state with the received insights. The constructor
+// then creates a function that other controllers can use to send insights to this controller, using the informerMsg
+// structure as a data transfer object.
+type updateStatusController struct {
+	configMaps corev1client.ConfigMapInterface
+
+	// statusApi is the desired state of the status API ConfigMap. It is updated when new insights are received.
+	// Any access to the struct should be done with the lock held.
+	statusApi struct {
+		sync.Mutex
+		cm *corev1.ConfigMap
+
+		// processed is the number of insights processed, used for testing
+		processed int
+	}
+
+	recorder events.Recorder
+}
+
+// newUpdateStatusController creates a new update status controller and returns it. The second return value is a function
+// the other controllers should use to send insights to this controller.
+func newUpdateStatusController(
+	ctx context.Context,
+	coreClient kubeclient.Interface,
+	coreInformers kubeinformers.SharedInformerFactory,
+	recorder events.Recorder,
+) (factory.Controller, sendInsightFn) {
+	uscRecorder := recorder.WithComponentSuffix("update-status-controller")
+
+	c := &updateStatusController{
+		configMaps: coreClient.CoreV1().ConfigMaps(uscNamespace),
+		recorder:   uscRecorder,
+	}
+
+	cmInformer := coreInformers.Core().V1().ConfigMaps().Informer()
+	controller := factory.New().
+		// call sync every 5 minutes or on CM events in the openshift-cluster-version namespace
+		WithSync(c.sync).ResyncEvery(5*time.Minute).
+		WithFilteredEventsInformersQueueKeysFunc(cmNameKey, nsFilter(uscNamespace), cmInformer).
+		ToController("UpdateStatusController", c.recorder)
+
+	// Spawn the insight collecting goroutine
+	sendInsight := c.startCollectingInsights(ctx)
+
+	return controller, sendInsight
+}
+
+// startCollectingInsights starts a goroutine that listens for insights from informers and updates internal status
+// API-mimicking ConfigMap.
+func (c *updateStatusController) startCollectingInsights(ctx context.Context) sendInsightFn {
+	fromInformers := make(chan informerMsg, 100)
```

Comment:
do we care about exhaustively processing everything that comes in?  Especially while we're in the "no preserved history" phase of this, I'd expect the sender to populate a locked map holding a single most-recently-pushed entry for that UID.  So the consumer wouldn't have to worry about things like "are there multiple updates for this UID in the queue?", the producer wouldn't have to worry about blocking while the consumer dug out of a hold, and nobody would have to worry about holding obsolete data in memory.

Reply:
Sending locked memory locations makes the controller slightly more coupled than I'd like at this point. My ambition was not to design the final inter-controller communication yet, more like make something that's very easy to understand (both code and while running) and eventually change. Right now I appreciate processing all messages, because it's easy to follow events across the system. Let's optimize later?

Reply:
But I'll leave a comment about the communication mechanism being intentionally simple.

Reply:
Partially addressed in faf8310f7f0bc4a61f699821b98ebd132b1639c9 (see https://github.com/openshift/cluster-version-operator/pull/1091#discussion_r1820945098)